### ISSUE TITLE
`SSHReader.read` hangs churning CPU when called after its channel is closed

### ISSUE THREAD

**Issue by Kentzo** [ronf/asyncssh#601]
If you take [my example from another issue](https://github.com/ronf/asyncssh/issues/599#issuecomment-1765695448) and change the condition inside the redirection coroutines from `not right_process.stdout.at_eof()` to `True`:

```python3
async def handle_client(left_process: asyncssh.SSHServerProcess) -> None:
    async with contextlib.AsyncExitStack() as stack:
        await stack.enter_async_context(left_process)

        conn = await stack.enter_async_context(
            asyncssh.connect(
                '192.168.2.1',
                username='test',
                password='test',
                preferred_auth=['password'],
                known_hosts=None,
                config=None
            )
        )
        right_process: asyncssh.SSHClientProcess = await stack.enter_async_context(
            conn.create_process(
                env=left_process.env,
                encoding=None,
                request_pty=True,
                term_type=left_process.term_type,
                term_size=left_process.term_size,
                term_modes=left_process.term_modes,
            )
        )

        async def redirect_stdin():
            n = left_process.channel.get_recv_window()
            while True:
                data = await left_process.stdin.read(n)
                right_process.stdin.write(data)

        async def redirect_stdout():
            n = right_process.channel.get_recv_window()
            while True:
                data = await right_process.stdout.read(n)
                left_process.stdout.write(data)

        redirect: asyncio.TaskGroup = await stack.enter_async_context(asyncio.TaskGroup())
        stack.callback(redirect.create_task(redirect_stdin()).cancel)
        stack.callback(redirect.create_task(redirect_stdout()).cancel)

        await right_process.channel.wait_closed()
```

Then the `handle_client` will never return if `right_process` gets to be closed first (e.g. via Ctrl-D).

Apparently it gets stuck because `SSHReader.read` never yields (hogging CPU at 100%), thus `SSHClientChannel.wait_closed()` never gets a chance to notice that the event was set and return.

The behavior I expect is to have asyncssh to raise `BrokenPipeError` whenever `SSHReader.read` or `SSHWriter.write` are called with the closed channel underneath.

**Comment by ronf** [comment]
It's not really safe to await on `wait_closed()` before calling `close()`. They are meant to go together, or you can use a context manager instead:

```python
async with asyncssh.connect() as conn:
    # do stuff with conn
```

In this case, `close()` will be called when exiting this block, along with awaiting on `wait_closed()`. This works for opening sessions as well, if you use `async with conn.create_process() as proc:` inside of this to start a session.

If you don't call close, you can end up with both sides waiting for the other to send EOF or to close the session/connection. In this specific case, you might get away with that if you want this code to keep running until the client/server initiates a close, but then you need to make sure you pass that through properly, which means looking for `read()` to return an empty str/bytes and calling `write_eof()` on the channel as soon as you get that. This should prevent the 100% CPU issue you saw, and in theory would allow each direction to close independently. You could get away with a `wait_closed()` in this case without calling `close()` yourself, but only if you properly pass through EOF so that the client/server will initiate the close.

You shouldn't need to manually call `cancel()` on the additional tasks you created in this case, as those tasks will exit once the EOF is received.

All that said, I'm still not clear on why you'd need code like this, other than as a debugging aid to capture the bytes sent back and forth. Was the client here an actual OpenSSH client, though, rather than your AsyncSSH client which didn't get the color sequences? Have you tried adding code in your AsyncSSH client to respond to the ESC Z with an appropriate sequence similar to what xterm returns?

**Comment by Kentzo** [comment]
I since updated the code to use `asyncio.wait` with `FIRST_COMPLETED` on `wait_closed` of both processes. If it’s still bad, what is the better approach to await process’s completion regardless of reason?

**Comment by ronf** [comment]
As long as you pass through EOF properly in your redirect_stdin() and redirect_stdout() using write_eof() when you get an empty string back from read(). This would be done only in one direction for each task (right_process.stdin.write_eof() from redirect_stdin and left_process.stdout.write_eof() from redirect_stdout), but it should trigger the client or server to do a close on the session, so you can safely do a wait_closed() without explicitly calling close() yourself, and you can do that in either order.

I would not recommend only waiting on one of the two -- if you try to exit the event loop while tasks are still running, you'll get warnings from asyncio.

Your other option would be to replace the write_eof() with close() calls on both processes from both redirect functions, but you'll still need to be looking for read() to return an empty string so you don't end up in a 100% CPU situation with read() returning immediately in a loop.

**Comment by Kentzo** [comment]
The actual code is something like:

```python3
    async with contextlib.AsyncExitStack() as stack:
        await stack.enter_async_context(left_process)

        conn = await stack.enter_async_context(
            asyncssh.connect(
                ...
            )
        )
        right_process: asyncssh.SSHClientProcess = await stack.enter_async_context(
            conn.create_process(
                ...
            )
        )

        await asyncio.wait([
            asyncio.create_task(left_process.wait_closed()),
            asyncio.create_task(right_process.wait_closed())
        ], return_when=asyncio.FIRST_COMPLETED)
```

Note how both processes are entered into the ExitStack. The idea is that whenever either of `wait_closed()` yields, the stack will take care of another.

I avoid handling EOF because it's not important for my process (i.e. there is no input / output that relies on it). My expectation is that it should suffice to terminate the remaining connection. Unless there is some other reason to be particularly careful with EOFs

---

My concern is that it would be unsafe to pass `SSHReader` / `SSHWriter` around, e.g. some other API that expects asyncio's `StreamReader` / `StreamWriter`, since their corresponding `read` and `write` do not raise when underlying channel is closed.

**Comment by ronf** [comment]
Ok - I can see how passing in the SSHProcess objects here will effecting trigger the equivalent of `wait_closed()` on them (and on the connection) when the AsyncExitStack's block exits. So, I think that may be ok here. However, if you don't handle EOF (empty string being returned by read()), one of your tasks may sit in an infinite loop once that session closes, and with the "cooperative" nature of the asyncio event loop, I'm not sure you'll ever make it back into the event loop to ever get a chance for this main block to exit.

If you pass through the EOF, that also has the benefit that it won't matter which task you end up waiting on, as one side closing will cause the other side to close as well, allowing you to wait for both `wait_closed()` calls to finish without worrying about the order.

That said, I think the only serious concern is the 100% CPU issue that EOF might trigger.
